README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# aae.pop.inference

<!-- badges: start -->

![R-CMD-check](https://github.com/aae-stats/aae.pop.inference/actions/workflows/check-standard.yaml/badge.svg)
[![Codecov test
coverage](https://codecov.io/github/aae-stats/aae.pop.inference/main/graph/badge.svg)](https://app.codecov.io/github/aae-stats/aae.pop.inference/)
<!-- badges: end -->

`aae.pop.inference` supports inference on population dynamics models
implemented in `aae.pop`. These models are relatively complex, making
standard inference methods (e.g. maximum likelihood, MCMC) challenging,
especially as the models become realistically complex (many population
classes, temporal variability). `aae.pop.inference` uses a sequential
approximate Bayesian computation approach to support fast inference on
these complex models.

## Installation

You can install the released version of aae.pop.inference from
[GitHub](https://github.com/aae-stats/aae.pop.inference) with:

``` r
remotes::install_github("aae-stats/aae.pop.inference")
```

## Example: estimating carrying capacity and density dependence

This example uses data simulated from a model as a target to estimate
the parameters of this model. This relatively simple example (estimating
parameters of a known model) illustrates how to perform inference on
these models but also highlights a case in which the target parameters
are highly correlated.

``` r
# some packages needed
library(aae.pop)
library(aae.pop.inference)
#> Loading required package: EasyABC
#> Loading required package: abc
#> Loading required package: abc.data
#> Loading required package: nnet
#> Loading required package: quantreg
#> Loading required package: SparseM
#> 
#> Attaching package: 'SparseM'
#> The following object is masked from 'package:base':
#> 
#>     backsolve
#> Loading required package: MASS
#> Loading required package: locfit
#> locfit 1.5-9.7    2023-01-02

# define a basic population model with aae.pop
nclass <- 5
popmat <- matrix(0, ncol = nclass, nrow = nclass)
popmat[transition(popmat)] <- c(0.3, 0.5, 0.7, 0.8)
popmat[reproduction(popmat)] <- c(0, 1, 4, 10)
dd <- density_dependence(
  funs = ricker(100, exclude = 1),
  masks = reproduction(popmat)
)
pop_fn <- dynamics(matrix = popmat, dd)

# simulate some data from this model to use as a target data set
obs <- simulate(
  pop_fn,
  nsim = 1000,
  args = list(density_dependence = list(theta = 0.4))
)

# define wrapper so we can simulate from the model based on
#   the inference parameters alone
popsim_fn <- function(par) {

  tmp <- update(
    pop_fn,
    density_dependence(
      funs = ricker(par[1], exclude = 1),
      masks = reproduction(popmat)
    )
  )

  sim <- simulate(tmp, nsim = 100, args = list(density_dependence = list(theta = par[2])))

  stat_abundance_trend(sim)

}

# calculate the target from the simulated data set
target <- stat_abundance_trend(obs)

# estimate parameters by comparing the simulation model to
#   the simulated data set with uniform priors on 
#   carrying capacity (U[50, 500]) and the theta parameter
#   of the Ricker density dependence (U[0.01, 1])
abc_est <- inference(
  model = popsim_fn,
  prior = list(
    c("unif", 50, 500),
    c("unif", 0.01, 1)
  ),
  target = target,
  nb_simul = 100,
  progress_bar = FALSE
)
```

Plotting the estimated posteriors for the two parameters suggests
relatively broad distributions with high uncertainty. However, a plot of
these two parameters against one another highlights their close
relationship, with a near perfect linear correlation between the two.

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" /><img src="man/figures/README-unnamed-chunk-2-2.png" width="100%" /><img src="man/figures/README-unnamed-chunk-2-3.png" width="100%" />

## Example: estimating vital rates

The `define()` function provides a helper to set up simulations when
inference is required for vital rates parameters only (the elements of
the population dynamics matrix). This function can be used to flexibly
perform inference on any or all elements in the vital rates matrix with
fixed or variable priors.

``` r
# define inference model for the transition and reproduction elements
popsim_vital <- define(
  x = pop_fn,
  masks = list(
    survival = aae.pop::transition(pop_fn$matrix),
    reproduction = aae.pop::reproduction(pop_fn$matrix)
  ),
  nsim = 100
)

# and run it (nb_simul too low and p_acc_min too high for a real example)
pars <- inference(
  model = popsim_vital$model,
  prior = popsim_vital$prior,
  target = popsim_vital$stat(obs),
  nb_simul = 100,   # real values should be closer to 1000
  progress_bar = FALSE,
  p_acc_min = 0.3   # default setting is 0.05
)

# plot the parameter estimates (one for each nonzero parameter)
plot(pars)
```

<img src="man/figures/README-define-example-1.png" width="100%" /><img src="man/figures/README-define-example-2.png" width="100%" />

## Issues

Please leave feedback, bug reports or feature requests at the GitHub
[issues page](https://github.com/aae-stats/aae.pop.inference/issues)


NEWS.md
# aae.pop.inference (development version)

## API changes

- Arguments to `simulate` are now passed as `sim_args` (a list) in `define`
- Dots argument in `define` passed to `stat` rather than `simulate`

# aae.pop.inference 0.1.0

## Features

- Basic implementation of SMC-ABC for population models created with aae.pop: `inference`
- Helpers to calculate summary statistics from replicated population trajectories: `stat_abundance_trend`, `stat_abundance_moment`
- Helper to create a simulation function for `inference` from an existing matrix population model defined with aae.pop: `define`



tests/testthat.R
library(testthat)
library(aae.pop)
library(aae.pop.inference)

test_check("aae.pop.inference")


tests/testthat/test_define.R
context("define")

# set a seed so the RNG aspects of this test don't cause random
#   failures
set.seed(251521)

# setup: define a pop model to simulate and use in ABC
nclass <- 5
popmat <- matrix(0, ncol = nclass, nrow = nclass)
popmat[transition(popmat)] <- c(0.3, 0.5, 0.7, 0.8)
popmat[reproduction(popmat)] <- c(0, 1, 4, 10)
dd <- density_dependence(
  funs = ricker(100, exclude = 1),
  masks = reproduction(popmat)
)
pop_fn <- dynamics(matrix = popmat, dd)

# simulate some data from this model to use as a target data set
obs <- simulate(
  pop_fn,
  nsim = 100,
  args = list(density_dependence = list(theta = 0.4))
)
obs_long <- simulate(
  pop_fn,
  nsim = 1000,
  args = list(density_dependence = list(theta = 0.4))
)

test_that("define works with fixed priors", {

  # define the inference model
  pop_inf <- define(x = pop_fn, sim_args = list(nsim = 5))

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(13L, ncol(pars$param))

})

test_that("define works with custom masks", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    sim_args = list(nsim = 5)
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

})

test_that("define works with custom priors", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    priors = list(
      survival = list(
        c("unif", 0.2, 0.4),
        c("unif", 0.3, 0.6),
        c("unif", 0.4, 0.9),
        c("unif", 0.5, 1)
      ),
      reproduction = list(
        c("lognormal", 0, 1),
        c("lognormal", 0, 1),
        c("lognormal", 0, 3),
        c("lognormal", 0, 3)
      )
    ),
    sim_args = list(nsim = 5)
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

})

test_that("define works accurately with custom priors", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    priors = list(
      survival = list(
        c("unif", 0.2, 0.4),
        c("unif", 0.3, 0.6),
        c("unif", 0.4, 0.9),
        c("unif", 0.5, 1)
      ),
      reproduction = list(
        c("lognormal", 0, 1),
        c("lognormal", 0, 1),
        c("lognormal", 0, 3),
        c("lognormal", 0, 3)
      )
    ),
    sim_args = list(
      nsim = 100,
      args = list(density_dependence = list(theta = 0.4))
    )
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs_long),
    nb_simul = 200,
    progress_bar = FALSE,
    p_acc_min = 0.05
  )

  # get true params
  target <- c(
    pop_fn$matrix[transition(pop_fn$matrix)],
    pop_fn$matrix[reproduction(pop_fn$matrix)]
  )

  # get modelled quantiles to see if true values are correlated
  value <- apply(pars$param, 2, quantile, p = 0.5)

  # just expect an output (not accurate) with correct number of pars and sims
  expect_gt(coef(lm(value ~ -1 + target)), 0.95)
  expect_lt(coef(lm(value ~ -1 + target)), 1.05)

})

test_that("define works with aae.pop args and options", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    sim_args = list(
      nsim = 5,
      args = list(density_dependence = list(theta = 0.4))
    )
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    sim_args = list(
      nsim = 5,
      args = list(density_dependence = list(theta = 0.4)),
      options = list(
        update = update_binomial_leslie,
        tidy_abundances = floor
      )
    )
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

})

test_that("define works correctly with different summary functions", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    stat = aae.pop.inference::stat_abundance_moment,
    sim_args = list(
      nsim = 5,
      args = list(density_dependence = list(theta = 0.4))
    )
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = pop_inf$stat(obs),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

})

test_that("define works when args are passed to summary functions", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    stat = aae.pop.inference::stat_abundance_moment,
    sim_args = list(
      nsim = 5,
      args = list(density_dependence = list(theta = 0.4))
    ),
    classes = list(c(1:3), c(3:5)),
    zscale = FALSE
  )

  # and run it
  pars <- aae.pop.inference::inference(
    model = pop_inf$model,
    prior = pop_inf$prior,
    target = do.call(pop_inf$stat, c(list(obs), pop_inf$stat_args)),
    nb_simul = 100,
    progress_bar = FALSE,
    p_acc_min = 0.3
  )

  # just expect an output (not accurate) with correct number of pars and sims
  expect_equal(8L, ncol(pars$param))

})

test_that("define errors informatively with incorrect priors", {

  # check survival as a list of incorrect dims
  expect_error(
    define(
      x = pop_fn,
      priors = list(
        survival = lapply(seq_len(25), function(i) c("unif", 1, 10))
      )
    ),
    "survival prior is a list with 25"
  )

  # check reproduction as a list of incorrect dims
  expect_error(
    define(
      x = pop_fn,
      priors = list(
        reproduction = lapply(seq_len(25), function(i) c("unif", 1, 10))
      )
    ),
    "reproduction prior is a list with 25"
  )

  # check survival as a vector of wrong class
  expect_error(
    define(
      x = pop_fn,
      priors = list(
        survival = c(1, 2, 3)
      )
    ),
    "priors must be character vectors or lists of character vectors"
  )

  # check reproduction as a vector of wrong class
  expect_error(
    define(
      x = pop_fn,
      priors = list(
        reproduction = c(1, 2, 3)
      )
    ),
    "priors must be character vectors or lists of character vectors"
  )

})

test_that("define errors informatively without dynamics input", {

  # check survival as a list of incorrect dims
  expect_error(
    define(x = rnorm(10)),
    "must be a dynamics object"
  )

})

test_that("print and is methods work correctly", {

  # define the inference model
  pop_inf <- define(
    x = pop_fn,
    masks = list(
      survival = aae.pop::transition(pop_fn$matrix),
      reproduction = aae.pop::reproduction(pop_fn$matrix)
    ),
    stat = aae.pop.inference::stat_abundance_moment,
    sim_args = list(
      nsim = 5,
      args = list(density_dependence = list(theta = 0.4))
    )
  )

  # check print returns an output
  expect_output(print(pop_inf), "Specified model and priors for use")

  # check the `is` method works
  expect_true(is.definition(pop_inf))

})


tests/testthat/test_inference.R
context("inference")

# set a seed so the RNG aspects of this test don't cause random
#   failures
set.seed(251521)

# setup: define a pop model to simulate and use in ABC
nclass <- 5
popmat <- matrix(0, ncol = nclass, nrow = nclass)
popmat[transition(popmat)] <- c(0.3, 0.5, 0.7, 0.8)
popmat[reproduction(popmat)] <- c(0, 1, 4, 10)
dd <- density_dependence(
  funs = ricker(100, exclude = 1),
  masks = reproduction(popmat)
)
pop_fn <- dynamics(matrix = popmat, dd)

# simulate some data from this model to use as a target data set
obs <- simulate(
  pop_fn,
  nsim = 100,
  args = list(density_dependence = list(theta = 0.4))
)

# repeat but with many more replicates
obs_long <- simulate(
  pop_fn,
  nsim = 1000,
  args = list(density_dependence = list(theta = 0.4))
)

# define wrapper for a quick simulation from parameters alone
popsim_fn <- function(par) {

  tmp <- update(
    pop_fn,
    aae.pop::density_dependence(
      funs = aae.pop::ricker(par[1], exclude = 1),
      masks = aae.pop::reproduction(popmat)
    )
  )

  sim <- simulate(
    tmp,
    nsim = 10,
    args = list(density_dependence = list(theta = par[2]))
  )

  stat_abundance_trend(sim)

}

# and repeat this but for more replicates
popsim_fn_long <- function(par) {

  tmp <- update(
    pop_fn,
    aae.pop::density_dependence(
      funs = aae.pop::ricker(par[1], exclude = 1),
      masks = aae.pop::reproduction(popmat)
    )
  )

  sim <- simulate(
    tmp,
    nsim = 100,
    args = list(density_dependence = list(theta = par[2]))
  )

  stat_abundance_trend(sim)

}

# calculate the target from the simulated data set
target <- stat_abundance_trend(obs)
target_long <- stat_abundance_trend(obs_long)

test_that("inference creates an ABC output", {

  # quick (inaccurate) inference
  inference_test <- inference(
    model = popsim_fn,
    prior = list(
      c("unif", 50, 500),
      c("unif", 0.01, 1)
    ),
    target = target,
    nb_simul = 10,
    progress_bar = FALSE
  )

  # quick inference using full function call
  easyabc_test <- ABC_sequential(
    method = "Lenormand",
    model = popsim_fn,
    prior = list(
      c("unif", 50, 500),
      c("unif", 0.01, 1)
    ),
    nb_simul = 10,
    summary_stat_target = target,
    progress_bar = FALSE
  )

  # check inference has the same output structure as an EasyABC call
  expect_equal(names(inference_test), names(easyabc_test))

})

test_that("inference generates an approximately correct output", {

  # longer (more accurate) inference
  inference_test <- inference(
    model = popsim_fn_long,
    prior = list(
      c("unif", 50, 500),
      c("unif", 0.01, 1)
    ),
    target = target_long,
    nb_simul = 100,
    progress_bar = FALSE
  )

  # linear regression of scaled parameter estimates should have
  #   slope near 1
  test_val <- coef(
    lm(
      scale(inference_test$param[, 1]) ~ -1 + scale(inference_test$param[, 2])
    )
  )

  # check inference has the same output structure as an EasyABC call
  expect_gt(test_val, 0.95)
  expect_lt(test_val, 1.05)

})

test_that("inference errors informatively", {

  # with incorrect model input
  expect_error(
    inference(
      model = target,
      prior = list(
        c("unif", 50, 500),
        c("unif", 0.01, 1)
      ),
      target = target,
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "must be a function"
  )

  # with incorrect prior class
  expect_error(
    inference(
      model = popsim_fn,
      prior = c(
        c("unif", 50, 500),
        c("unif", 0.01, 1)
      ),
      target = target,
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "must be a list"
  )

  # with incorrect prior values for a given distn
  expect_error(
    inference(
      model = popsim_fn,
      prior = list(
        c("unif", 50),
        c("unif", 0.01, 1)
      ),
      target = target,
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "at least one elemenet of prior has the wrong length"
  )

  # with incorrect prior values for a given distn
  expect_error(
    inference(
      model = popsim_fn,
      prior = list(
        c("exponential", 50, 10),
        c("unif", 0.01, 1)
      ),
      target = target,
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "at least one elemenet of prior has the wrong length"
  )

  # with un-implemented prior distribution
  expect_error(
    inference(
      model = popsim_fn,
      prior = list(
        c("labradoodle", 50, 500),
        c("unif", 0.01, 1)
      ),
      target = target,
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "prior contains unknown distribution"
  )

  # with mismatched model outputs and target
  expect_error(
    inference(
      model = popsim_fn,
      prior = list(
        c("unif", 50, 500),
        c("unif", 0.01, 1)
      ),
      target = c(target, 1),
      nb_simul = 10,
      progress_bar = FALSE
    ),
    "must return an output with the same length as target"
  )

})

test_that("S3 methods work correctly", {

  # quick (inaccurate) inference
  inference_test <- inference(
    model = popsim_fn,
    prior = list(
      c("unif", 50, 500),
      c("unif", 0.01, 1)
    ),
    target = target,
    nb_simul = 10,
    progress_bar = FALSE
  )

  # check `is` method
  expect_true(is.inference(inference_test))
  expect_false(is.inference(rnorm(5)))

  # check print returns an output
  expect_output(print(inference_test), "Outputs from sequential ABC")

  # check summary returns an output
  expect_output(summary(inference_test), "Inference on 2 model parameters")

  # test that an inference object can be plotted without warnings
  # TODO: consider more detailed test comparing expected to observed
  expect_silent(plot(inference_test))

})


R/inference.R
#' @name inference
#' @title Estimate parameters of a simulation model with sequential
#'   approximate Bayesian computation
#' @description Generic function to estimate parameters using sequential
#'   ABC, targeted specifically towards models implemented with the
#'   \code{\link[aae.pop:simulate]{aae.pop::simulate()}} function
#'   in the \code{aae.pop} package.
NULL

#' @rdname inference
#'
#' @export
#'
#' @importFrom EasyABC ABC_sequential
#' @importFrom stats runif rlnorm rnorm rexp quantile
#' @importFrom graphics hist par
#'
#' @param model a function to simulate a population dynamics object
#'   from a set of parameters. This function must return a
#'   vector of summary statistics with the same length as \code{target}
#' @param prior a list of prior distributions for each parameter, specified
#'   as described in
#'   \code{\link[EasyABC:ABC_sequential]{EasyABC::ABC_sequential()}}.
#'   Currently supported options are the uniform, normal, lognormal,
#'   and exponential distributions, and each list element should be a
#'   vector specifying the distribution ("unif", "normal", "lognormal",
#'   or "exponential"), followed by one (for the exponential distribution)
#'   or two parameters (all other distributions)
#' @param target values to be compared to model outputs simulated
#'   with \code{model}
#' @param method sequential ABC method to use for sampling, see
#'   \code{\link[EasyABC:ABC_sequential]{EasyABC::ABC_sequential()}} for
#'   details
#' @param x output from \code{inference}
#' @param y ignored, included for consistency with generic plot method
#' @param which the parameter to plot, defaults to all estimated parameters
#' @param object output from \code{inference}
#' @param \dots additional arguments passed to the sequential ABC sampler
#'   (see \code{\link[EasyABC:ABC_sequential]{EasyABC::ABC_sequential()}})
#'
#' @details Approximate Bayesian computation is a fast method for parameter
#'   inference in cases where model likelihoods are intractable (or at
#'   least difficult to compute). The \code{inference} function is a wrapper
#'   for the \code{\link[EasyABC:ABC_sequential]{EasyABC::ABC_sequential()}}
#'   function intended to simplify parameter inference for population
#'   dynamics models created with
#'   \code{\link[aae.pop:dynamics]{aae.pop::dynamics()}}.
#'
#'   Although standard MCMC algorithms are feasible for many of these models,
#'   the use of ABC allows fast and flexible specification of models based on
#'   existing simulation tools such as those in the \code{aae.pop} package.
#'
#' @examples
#' # use the tools in aae.pop to set up an example
#' library(aae.pop)
#'
#' # setup: define a pop model to simulate and use in ABC
#' nclass <- 5
#' popmat <- matrix(0, ncol = nclass, nrow = nclass)
#' popmat[transition(popmat)] <- c(0.3, 0.5, 0.7, 0.8)
#' popmat[reproduction(popmat)] <- c(0, 1, 4, 10)
#' dd <- density_dependence(
#'   funs = ricker(100, exclude = 1),
#'   masks = reproduction(popmat)
#' )
#' pop_fn <- dynamics(matrix = popmat, dd)
#'
#' # simulate some data from this model to use as a target data set
#' obs <- simulate(
#'   pop_fn,
#'   nsim = 100,
#'   args = list(density_dependence = list(theta = 0.4))
#' )
#'
#' # define wrapper for a simulation from K and DD parameters
#' popsim_fn <- function(par) {
#'
#'   tmp <- update(
#'     pop_fn,
#'     aae.pop::density_dependence(
#'       funs = aae.pop::ricker(par[1], exclude = 1),
#'       masks = aae.pop::reproduction(popmat)
#'     )
#'   )
#'
#'   sim <- simulate(
#'     tmp,
#'     nsim = 10,
#'     args = list(density_dependence = list(theta = par[2]))
#'   )
#'
#'   stat_abundance_trend(sim)
#'
#' }
#'
#' # calculate the target from the simulated data set
#' target <- stat_abundance_trend(obs)
#'
#' # and run it with uniform priors for both parameters (quick run,
#' #   nb_simul too low and p_acc_min too high for accurate results)
#' pars <- inference(
#'   model = popsim_fn,
#'   prior = list(
#'     c("unif", 50, 500),
#'     c("unif", 0.01, 1)
#'   ),
#'   target = target,
#'   nb_simul = 10,
#'   progress_bar = TRUE
#' )
#'
#' # plot estimated parameters (not expected to be accurate)
#' plot(pars)
inference <- function(
    model,
    prior,
    target,
    method = NULL,
    ...
) {

  # the simulation method must be a function and prior
  #   must be a list
  if (!is.function(model))
    stop("model must be a function", call. = FALSE)
  if (!is.list(prior))
    stop("prior must be a list", call. = FALSE)

  # and then check the prior is correctly specified for ABC_sequential
  check_prior(prior)

  # and the output from model should match the length of target
  prior_draw <- sapply(prior, sample_once)
  sim_test <- model(prior_draw)
  if (length(sim_test) != length(target)) {
    stop(
      "model must return an output with the same length as target",
      call. = FALSE
    )
  }

  # set baseline arguments and overwrite any that are set by user
  args <- list(
    model = model,
    prior = prior,
    summary_stat_target = target,
    method = "Lenormand",
    nb_simul = 1000,
    progress_bar = TRUE
  )
  args_dots <- list(...)
  args[names(args_dots)] <- args_dots

  # define and return inference output
  as_inference(do.call(EasyABC::ABC_sequential, args))

}

# internal function to check that the specified prior is defined
#   as required by EasyABC::ABC_sequential
check_prior <- function(x) {

  # check prior distn is implemented in ABC_sequential
  prior_distn <- sapply(x, function(x) x[1])
  if (!all(prior_distn %in% c("unif", "lognormal", "normal", "exponential"))) {
    stop(
      "prior contains unknown distribution; must be one of unif, normal, ",
      "lognormal, or exponential",
      call. = FALSE
    )
  }

  # check prior has the correct length for each type
  prior_ok <- TRUE
  for (i in seq_along(x)) {
    if (x[[i]][1] == "exponential") {
      args_ok <- length(x[[i]]) == 2 &
        !is.na(as.numeric(x[[i]][2]))
    } else {
      args_ok <- length(x[[i]]) == 3 &
        !any(is.na(as.numeric(x[[i]][2:3])))
    }
    prior_ok <- prior_ok & args_ok
  }

  # error if not OK
  if (!prior_ok) {
    stop(
      "at least one elemenet of prior has the wrong length or values; each ",
      " element should contain a distribution followed by one (for ",
      "exponential) or two numeric values (all other distributions)",
      call. = FALSE
    )
  }

}

# internal function to draw a single set of values from a prior defined
#   as described in EasyABC::ABC_sequential
sample_once <- function(x) {

  # specify a RNG for the appropriate distribution, defaulting
  #   to runif if not otherwise specified
  distn <- switch(
    x[1],
    "unif" = runif,
    "lognormal" = rlnorm,
    "normal" = rnorm,
    "exponential" = rexp,
    runif
  )

  # draw a single value from the relevant distribution
  do.call(distn, c(1, lapply(seq_along(x)[-1], function(i) as.numeric(x[i]))))

}

# S3 is method
#' @rdname inference
#' @export
# nolint start
is.inference <- function(x) {
  # nolint end
  inherits(x, "inference")
}

# S3 print method
#' @rdname inference
#' @export
# nolint start
print.inference <- function(x, ...) {
  # nolint end
  cat(
    paste0(
    "Outputs from sequential ABC inference ",
    "on population dynamics model\n"
    )
  )
}

# S3 summary method
#' @rdname inference
#' @export
# nolint start
summary.inference <- function(object, ...) {
  # nolint end

  # calculate quantiles of the estimated posterior distribution
  param_dist <- apply(
    object$param,
    2,
    quantile,
    pr = c(0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95)
  )
  colnames(param_dist) <- paste0("Parameter ", seq_len(ncol(object$param)))
  rownames(param_dist) <- paste0(
    c("5th", "10th", "25th", "50th", "75th", "90th", "95th"),
    " percentile"
  )

  # print these quantiles with info on compute time
  cat(paste0(
    "Inference on ", ncol(object$param),
    " model parameters took ", round(object$computime, 3),
    " seconds. Posterior distributions for each parameter ",
    " had the following quantiles:\n"
  ))
  print(round(param_dist, 4))

}

# S3 plot method
#' @rdname inference
#' @export
# nolint start
plot.inference <- function(x, y, ..., which = NULL) {
  # nolint end

  # grab plotting pars and reset at end
  old_mfrow <- par()$mfrow
  on.exit(par(mfrow = old_mfrow))

  # how many parameters are there?
  nparam <- ncol(x$param)

  # plot all if a subset isn't specified
  if (is.null(which))
    which <- seq_len(nparam)

  # set some basic arguments to keep the plot clean
  args <- list(
    main = "",
    xlab = "Parameter",
    las = 1
  )

  # overwrite with `...` if required
  arg_list <- list(...)
  args[names(arg_list)] <- arg_list

  # work out plotting dimensions
  sqrt_npar <- sqrt(length(which))
  ncol <- floor(sqrt_npar)
  nrow <- ceiling(sqrt_npar)
  par(mfrow = c(nrow, ncol))
  for (i in which)
    do.call(hist, c(list(x$param[, i]), args))

}

# internal function: set inference class
as_inference <- function(x) {
  as_class(x, name = "inference", type = "list")
}
